tsdb/chunkenc/value.go

```diff
@@ -16,23 +16,44 @@ package chunkenc
 import (
 	"bytes"
 	"encoding/binary"
+	"io"
+
+	"github.com/klauspost/compress/zstd"
 )
 
+// zstdFrameMagic is the magic beginning of all zstd compressed frames.
+var zstdFrameMagic = []byte{0x28, 0xb5, 0x2f, 0xfd}
+
 // valueChunk needs everything the ByteChunk does except timestamps.
 // The ValueIterator should just return []byte like the timestampChunk just returns timestamps.
-// The appender should just add the []byte as they are passed, no compression etc. (yet).
 
 type valueChunk struct {
-	b   []byte
-	num uint16
+	compressed []byte // only read once into b to decompress
+	b          []byte // appended to by Appender and decompressed to by Iterator if compressed before
+	num        uint16
 }
 
 func newValueChunk() *valueChunk {
 	return &valueChunk{b: make([]byte, 0, 5000)}
 }
 
 func (c *valueChunk) Bytes() []byte {
-	return c.b
+	// All samples of the chunk are uncompressed in c.b
+	// Before we return these []byte we compress them with zstd.
+	compressed := &bytes.Buffer{}
+	encoder, err := zstd.NewWriter(compressed, zstd.WithEncoderLevel(zstd.SpeedFastest))
+	if err != nil {
+		panic(err) // TODO don't panic
```

Comment:
did you mean to leave these TODO's in here?

Reply:
Yeah, I not sure myself what to do in these cases.
I can remove the comments and leave the panics, or we ignore the errors...
Either way not perfect.

Reply:
Log the error and return nil?

Reply:
Nope. Nothing at most `log.Println` from stdlib...

Reply:
I guess the question boils down to. Do we panic and stop the entire system if we cannot decompress one valueChunk or do we simply not care and return nil?
I'm probably leaning to the latter and that would mean ignoring errors for the most part.

Reply:
This is usually a case for an error iterator.

Reply:
Good idea. I create the valueIterator first and the set the error on that iterator in case it happens during the decompression. That way by the time `Next()` is called it returns `false` and the error gets surfaced through `it.Err()`.